### DISCUSSION TITLE
How to address array elements?

### DISCUSSION THREAD

**Discussion by tomdotorg** [bellrichm/WeeWX-MQTTSubscribe#148] [Q&A] [answered]
Hi, I am trying to store some values that are in arrays in the incoming json. Here is an example:

```
"xst": [
    35,
    34,
    32,
    38
  ],

```
In this case, the 35 should map to soilTemp1, and so on.

I can't see how that is done and would appreciate any pointers to docs. If there aren't any, I am happy to add to the complete json example.

**Comment by bellrichm** [comment]
Unfortunately this format is currently not supported. I am thinking the configuration would look like this.
```
    [[topics]]
        # The first topic to subscribe to. 
        [[[first/topic]]]
            # The incoming field name from MQTT.
            [[[[xst]]]]
                [[[[[subfields]]]]]
                     [[[[[[soilTemp1]]]]]]

                     [[[[[[soilTemp2]]]]]]
```
I haven't touched the json parser in a long time, so I'm not sure how long it will take to implement.  I'll keep you updated on the progress.  
rich

**Reply by tomdotorg** [reply]
Oh that's a bummer. I use those sensors so lack of support for that format means I will wait for it before making the switch over to this architecture. I do not have the Python skills to do it and submit a PR for it or I certainly would.

To be clear, the subfields would address elements in the array in order? That makes sense to me.

**Reply by bellrichm** [reply]
That’s the current thought. 
Also, all ‘field’ options would be supported ‘under’ subfields. subfields would inherit fields options and could override them. 
i’ve god a pretty good plan on how to implement this. Let me know if you are interested in any alpha or beta code. alpha code would pass current tests and pass the eye test for the new function. beta would pass basic tests for new function, but not all edge cases. 
rich

**Comment by tomdotorg** [comment]
I would *love* to help vet alpha, then beta. I have a nice side by side
install of weewx talking to a Wifi Logger 2 via WeatherLink and one that
pushes MQTT - they are both listening to the same ISS.

--
Tom Mitchell
***@***.***
Real Time Beverly, MA Weather
http://weather.tom.org

"The whole problem with the world
  is that fools and fanatics are
  always so certain of themselves,
  and wiser people so full of doubts."
  - Bertrand Russell

**Comment by bellrichm** [comment]
I've got something running...
With this payload
```
{
  "foo": 9, 
  "xst": [
    35, 
    34, 
    32, 
    38
  ]
}
```
And this configuration
```
    [[topics]]
        [[[weather/loop]]]
            [[[[foo]]]]
            [[[[xst]]]]
                [[[[[subfields]]]]]
                     [[[[[[soilTemp1]]]]]]
                     [[[[[[soilTemp2]]]]]]           
```
I got this output
```
2022-04-01 13:34:10 DEBUG (Driver) MessageCallbackProvider data-> incoming topic: weather/loop, QOS: 0, retain: 0, payload: b'{"foo": 9, "xst": [ 35, 34, 32, 38 ]}'
2022-04-01 13:34:16 DEBUG (Driver) TopicManager data-> incoming weather/loop: 'foo': '9.0', 'soilTemp1': '35.0', 'soilTemp2': '34.0'
2022-04-01 13:34:16 DEBUG (Driver) TopicManager data-> outgoing weather/loop: 'dateTime': '1648834456.2551324', 'foo': '9.0', 'soilTemp1': '35.0', 'soilTemp2': '34.0', 'usUnits': '1'
2022-04-01 13:34:18 DEBUG (Driver) data-> final loop packet is weather/loop 2022-04-01 13:34:16 EDT (1648834456): 'dateTime': '1648834456.2551324', 'foo': '9.0', 'soilTemp1': '35.0', 'soilTemp2': '34.0', 'usUnits': '1'

I'm working in the development branch. Are you comfortable pulling direct from there? Or do you want me to package up a release?
rich
```

**Reply by tomdotorg** [reply]
That's so great. I will grab the dev branch in the next few days. Should I report findings to this discussion or do you want issues? Thanks, Rich.

**Reply by bellrichm** [reply]
For now just comment here. If it gets out of hand we can reevaluate. 
There is no error checking. That is my next thing to work on, but life might slow the progress...

**Reply by tomdotorg** [reply]
I guess I misspoke. I am comfortable checking out the dev branch, but packaging that into something wee_extension doesn't complain about is something I haven't done before. I don't need a release, but a quick step from the zip I download from here to a valid weewx extension would be appreciated.. Sorry.

**Reply by roe-dl** [reply]
I would guess that way:
`wget -O weewx-MQTTSubscribe.zip https://github.com/bellrichm/WeeWX-MQTTSubscribe/archive/development.zip`

**Reply by bellrichm** [reply]
I’ve never done it, but you should be able to just run wee_extension against the zip file or if cloned, the WeeWX-MQTTSubscribe directory.

**Comment by tomdotorg** [comment]
That worked. So it's running and I will fold in the array elements to test
later, probably not until tomorrow.

--
Tom Mitchell
***@***.***
Real Time Beverly, MA Weather
http://weather.tom.org

"The whole problem with the world
  is that fools and fanatics are
  always so certain of themselves,
  and wiser people so full of doubts."
  - Bertrand Russell

**Comment by tomdotorg** [comment]
Quick update: I have this installed and configured this on two sites:

https://weewx-dev-classic.tom.org/ - this depends on my laptop's connection to the web (where mqtt is running) as well as a local database. Very much pure experimentation and local development. Also, no use of the soil temp or moisture sensors yet.

These two sites run against the same station/topics. One is seasons skin and one is Belchertown. Both run in the cloud on Kubernetes/Docker. More stable and where I will test. Uses soilTemp1-4 and soilMoist3.

https://vtwx-dev-classic.tom.org/ - Seasons skin
https://vtwx-dev.tom.org/ - Belchertown

**Reply by bellrichm** [reply]
Cool!

**Comment by bellrichm** [comment]
I have added a check that the length of the array is the same as the number of subfields.
Next up, cleanup some the configuration logging.

**Comment by bellrichm** [comment, marked as answer]
Subfields should now inherit configuration options.
I’d consider code now beta. I don’t plan on any updates unless you find something that doesn’t work or doesn’t make sense.

**Comment by tomdotorg** [comment]
OK, I'll get a PR ready for ya. Assuming a well-commented weewx.conf is
what you're looking for in the exampleConfigs directory is what you're
looking for or should I do a subdirectory under it where I can document how
to config the WFL and also weewx?

--
Tom Mitchell
***@***.***
Real Time Beverly, MA Weather
http://weather.tom.org

"The whole problem with the world
  is that fools and fanatics are
  always so certain of themselves,
  and wiser people so full of doubts."
  - Bertrand Russell

**Reply by bellrichm** [reply]
If possible I’d prefer just the one file. But if it makes more sense to have additional files, I think I’d like to keep the MQTTSubscribe config in the top directory and reference the subdirectory files.
I’m pretty flexible, you know better than me on what would make the most sense.

**Comment by tomdotorg** [comment]
Rich, is this working now, including soil moisture?

**Reply by bellrichm** [reply]
Oops, yes [v2.2.2](https://github.com/bellrichm/WeeWX-MQTTSubscribe/releases/tag/v2.2.2) should work.
In v2.2.1 I had forgotten to increment past the subfields being ignired.

**Reply by tomdotorg** [reply]
Hmm, I am pretty sure I submitted my complete config a few months ago #152 - were you looking for something more/different because I am happy to help?

**Reply by bellrichm** [reply]
You did.   But I got an email from GitHub this morning with that comment…
I was confused, but thought maybe you wanted to add/update. 
As far as I’m concerned, we are all set

**Comment by tomdotorg** [comment]
Yeah I got that one out of the blue too. I appreciate your work and will be
cutting some weewx over. Please let me know how I can do more to help.

**Reply by bellrichm** [reply]
Appreciate the kind words. At this time I can't think of anything that I need, but I will reach out if I do.
